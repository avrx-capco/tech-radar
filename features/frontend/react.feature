Feature: React content page

Background: Navigate to the home page
    Given the user navigates to "http://localhost:3000/"

Scenario: React content appears as expected
    When the user clicks "View Our Tech Radar"
    And the user clicks the hamburger menu
    And the user navigates to "Frontend" via the hamburger menu
    And the user navigates to category "React"
    And the user should see "React is the world’s most popular JavaScript library that uses components to change content of a single-page Mobile and Web application without refreshing."
    And the user should see "Managing states is one of the biggest challenges in large web applications. Using React alone may not be sufficient to handle such complexity. We need a way to share data across components by creating a central store that holds all states of your application."
    And the user should see "The first purpose of testing is to prevent bugs from releasing to production. Testing ensures the features are functioning correctly for complex components and modular applications and is required for the effective performance of a software application or product making it more robust and less prone to error."
    And the user should see "A React, TypeScript web application that provides a more readable format of our Engineering Competency Framework to help explain performance expectations. Unit tests are written in Jest and React Testing Library, E2E tests are written in Cypress and are automated via GitHub Actions."
    And the user clicks accordion "Redux"
    And the user can see accordion "Redux" description "Redux provides a central store for states that each component can access. Main building parts: store, actions, reducers and optional middleware."
    And the user clicks accordion "ContextAPI"
    And the user can see accordion "ContextAPI" description "Context provides a way to pass data through the component tree without having to pass props down manually at every level and helps to avoid the prop-drilling problem. It has 3 main building parts: context, provider and consumer."
    And the user clicks accordion "MaterialUI"
    And the user can see accordion "MaterialUI" description "Google’s Material Design is a popular React UI library with responsive animations and it has been adopted by the wider design community."
    And the user clicks accordion "SemanticUI"
    And the user can see accordion "SemanticUI" description "SemanticUI contains pre-built semantic components that helps create responsive layouts."
    And the user clicks accordion "Ant Design of React"
    And the user can see accordion "Ant Design of React" description "Ant Design of React contains a set of high-quality components and demos for building rich, interactive user interfaces."
    And the user clicks accordion "CSS and SASS"
    And the user can see accordion "CSS and SASS" description "This involves using separate stylesheets like our conventional way of styling our HTML websites either with CSS or a CSS preprocessor called SASS."
    And the user clicks accordion "BEM - Block Element Modifier"
    And the user can see accordion "BEM - Block Element Modifier" description "It's a CSS naming convention for writing cleaner and more readable CSS classes. BEM aims to write independent CSS blocks increasing reusability."
    And the user clicks accordion "CSS Modules"
    And the user can see accordion "CSS Modules" description "A CSS Module is a CSS file in which all class names and animation names are scoped locally by default. The key words here are scoped locally. With this, your CSS class names become similar to local variables in JavaScript. It goes into the compiler, and CSS comes out the other side."
    And the user clicks accordion "Styled Components"
    And the user can see accordion "Styled Components" description "styled-components is a library for React and React Native that allows you to use component-level styles in your application that are written with a mixture of JavaScript and CSS using a technique called CSS-in-JS."
    And the user clicks accordion "Jest"
    And the user can see accordion "Jest" description "Jest is an open-source testing framework. It comes with built-in mocking and assertion libraries, which also runs your tests concurrently in parallel, providing faster test execution."
    And the user clicks accordion "Enzyme"
    And the user can see accordion "Enzyme" description "Enzyme is an open-source JavaScript testing utility by Airbnb and needs a testing library to provide a foundation for the tests it generates. Using Jest and Enzyme together results in cleaner testing code that’s also easier to debug when a test breaks. Jest as a test runner and assertion library, then use Enzyme to build the tests for their UI."
    And the user clicks accordion "React Testing Library"
    And the user can see accordion "React Testing Library" description "The React Testing Library is a very lightweight solution for testing React components. It provides light utility functions on top of react-dom and react-dom/test-utils, in a way that encourages better testing practices."
    And the user clicks accordion "Cypress"
    And the user can see accordion "Cypress" description "Cypress is a JavaScript based end-to-end testing framework that doesn't use Selenium at all. It is built on top of Mocha, which is a feature-rich JavaScript test framework, making asynchronous testing simple. Cypress also uses a BDD/TDD assertion library and a browser that can be paired with any JavaScript testing framework."