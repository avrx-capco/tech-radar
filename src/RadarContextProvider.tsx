import { createContext, useState, FC } from 'react';

export type RadarContextType = {
  category: string;
  technology: string;

  setCategory: (newCategory: string) => void;
  setTechnology: (newComptenecy: string) => void;
};

export const RadarContext = createContext<RadarContextType>({
  category: '',
  technology: '',
  setCategory: (name: string) => {},
  setTechnology: (name: string) => {},
});

export const RadarContextProvider: FC = ({ children }) => {
  const [category, setCategory] = useState('');
  const [technology, setTechnology] = useState('');
  const initialValue = {
    category,
    technology,
    setCategory,
    setTechnology,
  };

  return (
    <RadarContext.Provider value={initialValue}>
      {children}
    </RadarContext.Provider>
  );
};
