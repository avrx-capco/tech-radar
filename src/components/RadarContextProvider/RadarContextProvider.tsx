/* eslint-disable @typescript-eslint/no-empty-function */
import { createContext, useState, FC } from "react";

export type RadarContextType = {
  category: string;
  technology: string;
  setCategory: (newCategory: string) => void;
  setTechnology: (newComptenecy: string) => void;
};

export const RadarContext = createContext<RadarContextType>({
  category: "",
  technology: "",
  setCategory: () => {},
  setTechnology: () => {},
});

export const RadarContextProvider: FC = ({ children }) => {
  const [category, setCategory] = useState("");
  const [technology, setTechnology] = useState("");
  const initialValue = {
    category,
    technology,
    setCategory,
    setTechnology,
  };

  return (
    <RadarContext.Provider value={initialValue}>
      {children}
    </RadarContext.Provider>
  );
};
