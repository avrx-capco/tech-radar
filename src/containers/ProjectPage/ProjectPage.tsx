import { useContext, useEffect, useState } from "react";
import {
  RadarContextType,
  RadarContext,
} from "components/RadarContextProvider/RadarContextProvider";
import ProjectItem from "components/ProjectItem/ProjectItem";
import { icons } from "data/data";
import ClientProjects from "data/projects/ClientProjects";
import { Container } from "@mui/material";
import { BackButton } from "components/BackLink/BackLink";
import Tabs from "components/Tabs/Tabs";
import { Wrapper, ContentBody, Title } from "./styles/";
import Axios from "axios";

/**
 * Project is an object with a project property that is a string, a clientName property that is a
 * string, a description property that is a string, and a
 * technologies property that is an array of strings.
 * @property {string} project - The name of the project.
 * @property {string} clientName - The name of the client.
 * @property {string} description - A short description of the project.
 * @property technologies - Array<string>;
 */
type Project = {
  project: string;
  clientName: string;
  description: string;
  technologies: Array<string>;
};

/* ProjectPage component that is using the useContext hook to get the technology from the
RadarContextProvider. It is also using the useState hook to set the state of the component. It is
also using the useEffect hook to fetch the client projects and the open source projects. */
const ProjectPage = () => {
  const { technology } = useContext<RadarContextType>(RadarContext);

  const [clientProjects, setClientProjects] = useState<Project[]>([]);
  const [iconRef, setIconRef] = useState<string>("");
  const [ossProjects, setOssProjects] = useState([]);

  /**
   * It fetches data from the Github API, filters the data based on the technology name, and then sets
   * the state of the component
   * @param {string} techName - string - this is the name of the technology you want to search for.
   */
  const fetchOssProject = async (techName: string) => {
    const response = await Axios.get(
      "https://api.github.com/orgs/capcodigital/repos"
    );
    const data = await response.data;
    const results = data.filter(({ topics }: any) => topics.includes(techName));

    const picked =
      results &&
      results.map(({ name, topics, html_url, description }: any) => ({
        project: name.replaceAll("-", " "),
        technologies: topics.flatMap((technology: any) =>
          technology.replaceAll("-", " ")
        ),
        githubLink: html_url,
        description: description,
      }));

    setOssProjects(picked);
  };

  /**
   * If the technology is not null, filter the ClientProjects array by the technology, otherwise filter
   * the ClientProjects array by the techName.
   * @param {string} techName - string - The name of the technology that was clicked on.
   */
  const fetchClientProjects = (techName: string) => {
    let results: Project[] = [];
    const splitTechName: string = techName?.toLowerCase().replace(/-/g, " ");
    const iconResult = icons.filter(
      (icon: any) => icon.name.toLowerCase() === splitTechName
    )[0];
    const ref = iconResult ? iconResult.link : "";
    setIconRef(ref);

    results = technology
      ? ClientProjects.filter(({ technologies }) =>
          technologies.includes(technology)
        )
      : ClientProjects.filter(({ technologies }) =>
          technologies.some((element) => {
            return element.toLowerCase() === splitTechName;
          })
        );

    setClientProjects(results);
  };

  /* Fetching the client projects and the open source projects. */
  useEffect(() => {
    const url = window.location.pathname.split("/");
    fetchClientProjects(url[3]);
    fetchOssProject(url[3]);
  }, []);

  /* Returning the JSX that is rendered on the page. */
  return (
    <Wrapper>
      <Container maxWidth="md">
        <BackButton />
        <Title>Projects</Title>
        <svg className="tech-icon" viewBox={"0 0 80 80"}>
          <circle cx={40} cy={40} r={40} fill={"white"} />
          <image x={15} y={15} href={iconRef} height={50} width={50} />
        </svg>
        <ContentBody>
          {clientProjects.length + ossProjects.length > 0 ? (
            <Tabs
              ossProjectCount={ossProjects.length}
              clientProjectCount={clientProjects.length}
              panelOne={
                clientProjects.length > 0 ? (
                  <ProjectItem data={clientProjects} />
                ) : (
                  <p>Projects not available</p>
                )
              }
              panelTwo={
                ossProjects.length > 0 && <ProjectItem data={ossProjects} />
              }
            />
          ) : (
            <p>Projects not available</p>
          )}
        </ContentBody>
      </Container>
    </Wrapper>
  );
};

export default ProjectPage;
