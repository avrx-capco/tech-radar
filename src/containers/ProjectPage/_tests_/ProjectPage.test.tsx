import * as React from "react";
import { render } from "@testing-library/react";
import { RadarContext } from "components/RadarContextProvider/RadarContextProvider";
import ProjectPage from "../ProjectPage";
import { MemoryRouter } from "react-router-dom";
import mockAxios from "jest-mock-axios";
import { act } from "react-dom/test-utils";

const mockValue = {
  category: "mobile",
  technology: "Swift",
  setCategory: jest.fn(),
  setTechnology: jest.fn(),
};

/* Mocking the RadarContextProvider component. */
jest.doMock("components/RadarContextProvider/RadarContextProvider", () => {
  return {
    __esModule: true,
    default: React.createContext,
  };
});

window.scroll = jest.fn();

describe("ProjectPage", () => {
  beforeEach(() => {
    mockAxios.get.mockResolvedValue({ data: [] });
  });
  afterEach(() => {
    jest.restoreAllMocks();
    mockAxios.reset();
  });

  it("when techology is React it should render ProjectPage component", async () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let renderedComponent: any;
    await act(async () => {
      renderedComponent = render(
        <MemoryRouter>
          <RadarContext.Provider value={mockValue}>
            <ProjectPage />
          </RadarContext.Provider>
        </MemoryRouter>
      );
    });
    expect.addSnapshotSerializer({
      // Identify if there is any dynamic prop created by MUI (Material UI).
      test: (val: any) =>
        val && typeof val === "string" && val.indexOf("mui-") >= 0,
      print: (val: any) => {
        // Replace any dynamic prop value created by MUI with a constant value for avoiding Snapshot mismatching.
        let str = val;
        str = str.replace(/mui-.*/gm, "mui-00000");

        return `"${str}"`;
      },
    });
    expect(mockAxios.get).toHaveBeenCalledWith(
      "https://api.github.com/orgs/capcodigital/repos"
    );
    expect(renderedComponent).toMatchSnapshot();
  });
});
