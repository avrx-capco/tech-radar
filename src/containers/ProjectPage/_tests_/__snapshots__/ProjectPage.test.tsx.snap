// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectPage when techology is React it should render ProjectPage component 1`] = `
<div
  class="styles__Wrapper-sc-azja3a-0 bzpueU"
>
  <div
    class="MuiContainer-root MuiContainer-maxWidthMd"
  >
    <button
      class="styles__StyledBackButton-sc-1t7dby2-1 irAvQb"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
        />
      </svg>
      Back
    </button>
    <div
      class="styles__Title-sc-azja3a-2 cpmuRv"
    >
      Projects
    </div>
    <svg
      class="tech-icon"
      viewBox="0 0 80 80"
    >
      <circle
        cx="40"
        cy="40"
        fill="white"
        r="40"
      />
      <image
        height="50"
        href=""
        width="50"
        x="15"
        y="15"
      />
    </svg>
    <div
      class="styles__ContentBody-sc-azja3a-1 fCkuSW"
    >
      <div
        class="styles__StyledTabs-sc-gdc602-0 eZgCjI"
      >
        <div
          class="MuiBox-root css-1gsv261"
        >
          <div
            class="MuiTabs-root css-1ujnqem-MuiTabs-root"
          >
            <div
              class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
              style="overflow: hidden; margin-bottom: 0px;"
            >
              <div
                aria-label="Project tabs"
                class="MuiTabs-flexContainer MuiTabs-centered css-1wf8b0h-MuiTabs-flexContainer"
                role="tablist"
              >
                <button
                  aria-controls="mui-p-58332-P-1"
                  aria-selected="true"
                  class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                  id="mui-p-58332-T-1"
                  role="tab"
                  tabindex="0"
                  type="button"
                >
                  Client Projects (1)
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-controls="mui-p-58332-P-2"
                  aria-selected="false"
                  class="MuiButtonBase-root Mui-disabled MuiTab-root MuiTab-textColorPrimary Mui-disabled css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                  disabled=""
                  id="mui-p-58332-T-2"
                  role="tab"
                  tabindex="-1"
                  type="button"
                >
                  Capco OSS Projects (0)
                </button>
              </div>
              <span
                class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
                style="left: 0px; width: 0px;"
              />
            </div>
          </div>
        </div>
        <div
          aria-labelledby="mui-p-58332-T-1"
          class="MuiTabPanel-root css-13xfq8m-MuiTabPanel-root"
          id="mui-p-58332-P-1"
          role="tabpanel"
        >
          <div
            class="styles__StyledWrapper-sc-1pgv985-0 kuStYb"
            data-test-id="project-0"
          >
            <div>
              <div
                class="project-image-wrapper"
              >
                <img
                  alt="computer screen"
                  class="project-image"
                  src="CL3.png"
                />
              </div>
              <div
                class="row-wrapper"
              >
                <div
                  class="heading"
                >
                  Mobile Banking Remediation
                </div>
                <div
                  class="project-tag"
                >
                  Global Tier 1 Bank
                </div>
              </div>
              <div
                class="text"
              >
                Capco performed a holistic assessment of the clientâ€™s current mobile applications, architectures, development standards and implementation.  Our team set new standards, implemented coding design principles, deployed a new DevOps pipeline, implemented automation testing and restructured and refactored the current applications, reducing development time/cost which significantly increased customer app ratings.
              </div>
              <div
                class="heading"
              >
                Key Technologies
              </div>
              <div
                class="technologies"
              >
                <span>
                  <a
                    class="projectTech-link"
                    href="/technology/mobile/swift"
                  >
                    <svg
                      class="icon"
                      viewBox="0 0 80 80"
                    >
                      <circle
                        cx="40"
                        cy="40"
                        fill="white"
                        r="40"
                      />
                      <image
                        height="50"
                        href="https://img.icons8.com/fluent/452/swift.png"
                        width="50"
                        x="15"
                        y="15"
                      />
                    </svg>
                    <div>
                      Swift
                    </div>
                  </a>
                </span>
                <span>
                  <a
                    class="projectTech-link"
                    href="/technology/backend/kotlin"
                  >
                    <svg
                      class="icon"
                      viewBox="0 0 80 80"
                    >
                      <circle
                        cx="40"
                        cy="40"
                        fill="white"
                        r="40"
                      />
                      <image
                        height="50"
                        href="https://raw.githubusercontent.com/capcodigital/presentation-logos/master/tech-radar/backend/kotlin_square.png"
                        width="50"
                        x="15"
                        y="15"
                      />
                    </svg>
                    <div>
                      Kotlin
                    </div>
                  </a>
                </span>
                <span>
                  <a
                    class="projectTech-link"
                    href="/technology/devops/jenkins"
                  >
                    <svg
                      class="icon"
                      viewBox="0 0 80 80"
                    >
                      <circle
                        cx="40"
                        cy="40"
                        fill="white"
                        r="40"
                      />
                      <image
                        height="50"
                        href="https://raw.githubusercontent.com/capcodigital/presentation-logos/master/tech-radar/devops/jenkins_square.png"
                        width="50"
                        x="15"
                        y="15"
                      />
                    </svg>
                    <div>
                      Jenkins
                    </div>
                  </a>
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          aria-labelledby="mui-p-58332-T-2"
          class="MuiTabPanel-root css-13xfq8m-MuiTabPanel-root"
          hidden=""
          id="mui-p-58332-P-2"
          role="tabpanel"
        />
      </div>
    </div>
  </div>
</div>
`;
